server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java

```diff
@@ -464,6 +466,25 @@ private int submitCompactionTasks(
         }
       }
 
+      CompactionEngine compactionEngine = config.getEngine();
+      Map<String, Object> autoCompactionContext = newAutoCompactionContext(config.getTaskContext());
+      int numCurrentCompactionTasksAndSubtasks;
+      final String maxNumTasksContextParam = "maxNumTasks";
+
+      if (compactionEngine == CompactionEngine.MSQ) {
+        if (!autoCompactionContext.containsKey(maxNumTasksContextParam)) {
+          // Since MSQ needs all task slots for the calculated #tasks to be available upfront, allot all available
```

Comment:
While this is true, we can still impose a max value. For large clusters with a lot of compaction task slots, it doesn't make sense to take everything up for a single compaction job.

Let's start with some maximum value, say 6 or 8 workers. This can anyway be overridden by the `taskContext` provided in the `DatasourceCompactionConfig`.